mis_builder/models/mis_builder.py

```diff
@@ -0,0 +1,865 @@
+
+    def write(self, cr, uid, ids, vals, context=None):
+        # TODO: explain this
+        res = super(mis_report, self).write(
+            cr, uid, ids, vals, context=context)
+        mis_report_kpi_obj = self.pool.get('mis.report.kpi')
+        for report in self.browse(cr, uid, ids, context):
+            for idx, kpi in enumerate(report.kpi_ids):
+                mis_report_kpi_obj.write(
+                    cr, uid, [kpi.id], {'sequence': idx + 1}, context=context)
+        return res
+
+
+class mis_report_instance_period(orm.Model):
+
+    """ A MIS report instance has the logic to compute
+    a report template for a given date period.
+
+    Periods have a duration (day, week, fiscal period) and
+    are defined as an offset relative to a pivot date.
+    """
+
+    def _get_dates(self, cr, uid, ids, field_names, arg, context=None):
+        if isinstance(ids, (int, long)):
+            ids = [ids]
+        res = {}
+        for c in self.browse(cr, uid, ids, context=context):
+            d = parser.parse(c.report_instance_id.pivot_date)
+            if c.type == 'd':
+                date_from = d + timedelta(days=c.offset)
+                date_to = date_from + timedelta(days=c.duration - 1)
+                date_from = date_from.strftime(
+                    tools.DEFAULT_SERVER_DATE_FORMAT)
+                date_to = date_to.strftime(tools.DEFAULT_SERVER_DATE_FORMAT)
+                period_ids = None
+            elif c.type == 'w':
+                date_from = d - timedelta(d.weekday())
+                date_from = date_from + timedelta(days=c.offset * 7)
+                date_to = date_from + timedelta(days=(7 * c.duration) - 1)
+                date_from = date_from.strftime(
+                    tools.DEFAULT_SERVER_DATE_FORMAT)
+                date_to = date_to.strftime(tools.DEFAULT_SERVER_DATE_FORMAT)
+                period_ids = None
+            elif c.type == 'fp':
+                period_obj = self.pool['account.period']
+                all_period_ids = period_obj.search(
+                    cr, uid,
+                    [('special', '=', False),
+                     '|', ('company_id', '=', False),
+                     ('company_id', '=', c.company_id.id)],
+                    order='date_start',
+                    context=context)
+                current_period_ids = period_obj.search(
+                    cr, uid,
+                    [('special', '=', False),
+                     ('date_start', '<=', d),
+                     ('date_stop', '>=', d),
+                     '|', ('company_id', '=', False),
+                     ('company_id', '=', c.company_id.id)],
+                    context=context)
+                if not current_period_ids:
+                    raise orm.except_orm(_("Error!"),
+                                         _("No current fiscal period for %s")
+                                         % d)
```

Comment:
It is a bad idea to send exception in functional field !
Because you can not change anything if the value can not be computed.